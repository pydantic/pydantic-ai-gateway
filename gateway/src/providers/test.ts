import { DefaultProviderProxy } from './default'

export class TestProvider extends DefaultProviderProxy {
  providerId(): string {
    return 'openai'
  }

  apiFlavour(): string | undefined {
    return 'chat'
  }

  fetch(url: string): Promise<Response> {
    const data = {
      choices: [
        {
          finish_reason: 'stop',
          index: 0,
          message: { annotations: [], content: `request URL: ${url}`, refusal: null, role: 'assistant' },
        },
      ],
      created: Math.floor(Date.now() / 1000),
      id: 'chatcmpl-test',
      model: 'gpt-5',
      object: 'chat.completion',
      service_tier: 'default',
      system_fingerprint: null,
      usage: {
        prompt_tokens: 4560,
        completion_tokens: 1230,
        completion_tokens_details: {
          accepted_prediction_tokens: 0,
          audio_tokens: 0,
          reasoning_tokens: 0,
          rejected_prediction_tokens: 0,
        },
        prompt_tokens_details: { audio_tokens: 0, cached_tokens: 0 },
        total_tokens: 5790,
      },
    }
    const headers = { 'Content-Type': 'application/json', 'pydantic-ai-gateway': 'test' }
    return Promise.resolve(new Response(JSON.stringify(data), { status: 200, headers }))
  }
}
