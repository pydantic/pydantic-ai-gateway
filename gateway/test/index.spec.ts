import OpenAI from 'openai'
import Groq from 'groq-sdk'
import { env } from 'cloudflare:test'
import { describe, it, expect } from 'vitest'
import { test } from './conftest'

import { LimitDbD1 } from '@pydantic/ai-gateway'
import { IDS } from './worker'

describe('index', () => {
  test('responds with index html', async ({ gateway }) => {
    const response = await gateway.fetch('https://example.com')
    expect(response.status).toBe(200)
    expect(await response.text()).toMatchInlineSnapshot(
      `
      "▗▄▄▖  ▗▄▖ ▗▄▄▄▖ ▗▄▄▖
      ▐▌ ▐▌▐▌ ▐▌  █  ▐▌
      ▐▛▀▘ ▐▛▀▜▌  █  ▐▌▝▜▌
      ▐▌   ▐▌ ▐▌▗▄█▄▖▝▚▄▞▘

      Pydantic AI Gateway

      git sha: test
      GitHub: https://github.com/pydantic/pydantic-ai-gateway
      To connect, point your application at https://example.com/<provider-id>
      "
    `,
    )
    expect(response.headers.get('content-type')).toBe('text/plain; charset=utf-8')
  })
})

describe('invalid request', () => {
  test('401 on no auth header', async ({ gateway }) => {
    const response = await gateway.fetch('https://example.com/openai/gpt-5')
    const text = await response.text()
    expect(response.status, `got ${response.status} response: ${text}`).toBe(401)
    expect(text).toMatchInlineSnapshot(`"Unauthorized - Missing Authorization Header"`)
  })
  test('401 on unknown auth header', async ({ gateway }) => {
    const response = await gateway.fetch('https://example.com/openai/gpt-5', {
      headers: { Authorization: 'unknown-token' },
    })
    const text = await response.text()
    expect(response.status, `got ${response.status} response: ${text}`).toBe(401)
    expect(text).toMatchInlineSnapshot(`"Unauthorized - Key not found"`)
  })
  test('400 on unknown provider', async ({ gateway }) => {
    const response = await gateway.fetch('https://example.com/wrong/gpt-5', {
      headers: { Authorization: 'unknown-token' },
    })
    const text = await response.text()
    expect(response.status, `got ${response.status} response: ${text}`).toBe(400)
    expect(text).toMatchInlineSnapshot(
      `"Invalid provider 'wrong', should be one of groq, openai, google-vertex, anthropic"`,
    )
  })
})

describe('openai', () => {
  test('should call openai via gateway', async ({ gateway }) => {
    const { fetch, otelBatch } = gateway

    const client = new OpenAI({ apiKey: 'healthy', baseURL: 'https://example.com/openai', fetch })

    const completion = await client.chat.completions.create({
      model: 'gpt-5',
      messages: [
        { role: 'developer', content: 'You are a helpful assistant.' },
        { role: 'user', content: 'What is the capital of France?' },
      ],
    })

    expect(completion).toMatchSnapshot('llm')
    expect(otelBatch, 'otelBatch length not 1').toHaveLength(1)
    expect(JSON.parse(otelBatch[0]!).resourceSpans?.[0].scopeSpans?.[0].spans?.[0]?.attributes).toMatchSnapshot('span')

    const limitDb = new LimitDbD1(env.limitsDB)
    const teamStatus = await limitDb.spendStatus('team')
    expect(teamStatus).toEqual([
      { entityId: IDS.teamDefault, limit: 4, scope: 'monthly', scopeInterval: expect.any(Date), spend: 0.00013875 },
    ])
    const userStatus = await limitDb.spendStatus('user')
    expect(userStatus).toEqual([
      { entityId: IDS.userDefault, limit: 3, scope: 'weekly', scopeInterval: expect.any(Date), spend: 0.00013875 },
    ])
    const keyStatus = await limitDb.spendStatus('key')
    expect(keyStatus.sort((a, b) => a.limit - b.limit)).toEqual([
      { entityId: IDS.keyHealthy, limit: 1, scope: 'daily', scopeInterval: expect.any(Date), spend: 0.00013875 },
      { entityId: IDS.keyHealthy, limit: 2, scope: 'total', scopeInterval: null, spend: 0.00013875 },
    ])
  })
})

describe('groq', () => {
  test('should call groq via gateway', async ({ gateway }) => {
    const { fetch } = gateway
    const client = new Groq({ apiKey: 'healthy', baseURL: 'https://example.com/groq', fetch })

    const completion = await client.chat.completions.create({
      model: 'llama-3.3-70b-versatile',
      messages: [
        { role: 'developer', content: 'You are a helpful assistant.' },
        { role: 'user', content: 'What is the capital of France?' },
      ],
    })
    expect(completion).toMatchSnapshot('llm')
  })
})

describe('key status', () => {
  test('should not change key status if limit is not exceeded', async ({ gateway }) => {
    const { fetch } = gateway
    const client = new OpenAI({ apiKey: 'healthy', baseURL: 'https://example.com/test', fetch })
    await client.chat.completions.create({
      model: 'gpt-5',
      messages: [
        { role: 'developer', content: 'You are a helpful assistant.' },
        { role: 'user', content: 'Give me an essay on the history of the universe.' },
      ],
    })
    const allSpends = await env.limitsDB
      .prepare(
        `SELECT entityId, entityType, scope, round(spend, 3) spend, spendingLimit FROM spend order by spendingLimit`,
      )
      .run<{ entityId: number; entityType: number; scope: number; spend: string; spendingLimit: number }>()
    expect(allSpends.results).toMatchSnapshot('spend-table')
    const allKeyStatus = await env.limitsDB
      .prepare('SELECT count(*) as count FROM keyStatus')
      .first<{ count: number }>()
    expect(allKeyStatus?.count).toBe(0)
  })

  test('should block request if key is disabled', async ({ gateway }) => {
    const { fetch } = gateway

    const response = await fetch('https://example.com/openai/xxx', { headers: { Authorization: 'disabled' } })
    const text = await response.text()
    expect(response.status, `got response: ${response.status} ${text}`).toBe(403)
    expect(text).toMatchInlineSnapshot(`"Unauthorized - Key disabled"`)

    const spendCount = await env.limitsDB.prepare('SELECT count(*) count FROM spend').first<{ count: number }>()
    expect(spendCount?.count).toBe(0)
    const keyStatusCount = await env.limitsDB
      .prepare('SELECT count(*) count FROM keyStatus')
      .first<{ count: number }>()
    expect(keyStatusCount?.count).toBe(0)
  })

  test('should change key status if limit is exceeded', async ({ gateway }) => {
    const { fetch, disableEvents } = gateway

    expect(disableEvents).toEqual([])

    const client = new OpenAI({ apiKey: 'tiny-limit', baseURL: 'https://example.com/test', fetch })
    await client.chat.completions.create({
      model: 'gpt-5',
      messages: [
        { role: 'developer', content: 'You are a helpful assistant.' },
        { role: 'user', content: 'Give me an essay on the history of the universe.' },
      ],
    })

    const apiValue = await env.KV.get('apiKeyAuth:test:tiny-limit')
    expect(apiValue).toBeTypeOf('string')
    expect(JSON.parse(apiValue!)).toMatchSnapshot('kv-value')

    const allSpends1 = await env.limitsDB
      .prepare(
        `SELECT entityId, entityType, scope, round(spend, 3) spend, spendingLimit FROM spend order by spendingLimit`,
      )
      .run<{ entityId: number; entityType: number; scope: number; spend: string; spendingLimit: number }>()
    expect(allSpends1.results).toMatchInlineSnapshot(`
          [
            {
              "entityId": 5,
              "entityType": 3,
              "scope": 1,
              "spend": 0.018,
              "spendingLimit": 0.01,
            },
            {
              "entityId": 1,
              "entityType": 1,
              "scope": 3,
              "spend": 0.018,
              "spendingLimit": 4,
            },
          ]
        `)

    expect(disableEvents).toEqual([
      {
        id: IDS.keyTinyLimit,
        reason: 'limits exceeded: key-daily',
        newStatus: 'limit-exceeded',
        expirationTtl: expect.any(Number),
      },
    ])
    expect(disableEvents[0]!.expirationTtl).toBeGreaterThanOrEqual(0)
    expect(disableEvents[0]!.expirationTtl).toBeLessThanOrEqual(86400)

    const keyStatusQuery = await env.limitsDB
      .prepare("SELECT id, status, strftime('%s', expiresAt) - strftime('%s','now') as expiresAtDiff FROM keyStatus")
      .run<{ id: string; status: string; expiresAtDiff: number }>()
    expect(keyStatusQuery.results).toEqual([
      { id: IDS.keyTinyLimit, status: 'limit-exceeded', expiresAtDiff: expect.any(Number) },
    ])
    expect(Math.abs(keyStatusQuery.results[0]!.expiresAtDiff - disableEvents[0]!.expirationTtl!)).toBeLessThan(2)

    {
      const response = await fetch('https://example.com/openai/xxx', { headers: { Authorization: 'tiny-limit' } })
      const text = await response.text()
      expect(response.status, `got ${response.status} response: ${text}`).toBe(403)
      expect(text).toMatchInlineSnapshot(`"Unauthorized - Key limit-exceeded"`)
    }

    expect(disableEvents).toEqual([
      {
        id: IDS.keyTinyLimit,
        reason: 'limits exceeded: key-daily',
        newStatus: 'limit-exceeded',
        expirationTtl: expect.any(Number),
      },
    ])
  })
})

describe('LimitDbD1', () => {
  it('updates limit', async () => {
    const db = new LimitDbD1(env.limitsDB)
    await db.incrementSpend(
      [{ entityId: IDS.userDefault, entityType: 'user', scope: 'daily', scopeInterval: 123, limit: 2 }],
      1,
    )

    {
      const state = await env.limitsDB.prepare('SELECT * FROM spend').first()
      expect(state).toMatchInlineSnapshot(`
      {
        "entityId": 2,
        "entityType": 2,
        "scope": 1,
        "scopeInterval": 123,
        "spend": 1,
        "spendingLimit": 2,
      }
    `)
    }

    await db.updateUserLimits(IDS.userDefault, { daily: 3, weekly: 5 })

    {
      const state = await env.limitsDB.prepare('SELECT * FROM spend').first()
      expect(state).toMatchInlineSnapshot(`
      {
        "entityId": 2,
        "entityType": 2,
        "scope": 1,
        "scopeInterval": 123,
        "spend": 1,
        "spendingLimit": 3,
      }
    `)
    }
  })
})
