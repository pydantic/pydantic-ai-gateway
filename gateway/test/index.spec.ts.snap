// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`groq > should call groq via gateway > llm 1`] = `
{
  "choices": [
    {
      "finish_reason": "stop",
      "index": 0,
      "logprobs": null,
      "message": {
        "content": "The capital of France is Paris.",
        "role": "assistant",
      },
    },
  ],
  "created": 1757583859,
  "id": "chatcmpl-dc545ddf-4d56-41c5-b839-aa29388b76f5",
  "model": "llama-3.3-70b-versatile",
  "object": "chat.completion",
  "service_tier": "on_demand",
  "system_fingerprint": "fp_2ddfbb0da0",
  "usage": {
    "completion_time": 0.002192431,
    "completion_tokens": 8,
    "prompt_time": 0.010942305,
    "prompt_tokens": 42,
    "pydantic_ai_gateway": {
      "cost_estimate": 0.0000311,
    },
    "queue_time": 0.084393519,
    "total_time": 0.013134736,
    "total_tokens": 50,
  },
  "usage_breakdown": null,
  "x_groq": {
    "id": "req_01k4w2jqx2f31tmse2nw28za8v",
  },
}
`;

exports[`key status > should change key status if limit is exceeded > kv-value 1`] = `
{
  "id": 5,
  "key": "tiny-limit",
  "keySpendingLimitDaily": 0.01,
  "otelSettings": null,
  "providers": [
    {
      "baseUrl": "http://test.example.com/test",
      "credentials": "test",
      "injectCost": true,
      "providerID": "test",
    },
  ],
  "status": "limit-exceeded",
  "team": 1,
  "teamSpendingLimitMonthly": 4,
  "user": null,
}
`;

exports[`key status > should not change key status if limit is not exceeded > spend-table 1`] = `
[
  {
    "entityId": 3,
    "entityType": 3,
    "scope": 1,
    "spend": 0.018,
    "spendingLimit": 1,
  },
  {
    "entityId": 3,
    "entityType": 3,
    "scope": 4,
    "spend": 0.018,
    "spendingLimit": 2,
  },
  {
    "entityId": 2,
    "entityType": 2,
    "scope": 2,
    "spend": 0.018,
    "spendingLimit": 3,
  },
  {
    "entityId": 1,
    "entityType": 1,
    "scope": 3,
    "spend": 0.018,
    "spendingLimit": 4,
  },
]
`;

exports[`openai > should call openai via gateway > llm 1`] = `
{
  "choices": [
    {
      "finish_reason": "stop",
      "index": 0,
      "message": {
        "annotations": [],
        "content": "Paris.",
        "refusal": null,
        "role": "assistant",
      },
    },
  ],
  "created": 1757583776,
  "id": "chatcmpl-CEY7M6BwbBtavblK5zBwrXgbDdIaF",
  "model": "gpt-5-2025-08-07",
  "object": "chat.completion",
  "service_tier": "default",
  "system_fingerprint": null,
  "usage": {
    "completion_tokens": 11,
    "completion_tokens_details": {
      "accepted_prediction_tokens": 0,
      "audio_tokens": 0,
      "reasoning_tokens": 0,
      "rejected_prediction_tokens": 0,
    },
    "prompt_tokens": 23,
    "prompt_tokens_details": {
      "audio_tokens": 0,
      "cached_tokens": 0,
    },
    "pydantic_ai_gateway": {
      "cost_estimate": 0.00013875,
    },
    "total_tokens": 34,
  },
}
`;

exports[`openai > should call openai via gateway > span 1`] = `
[
  {
    "key": "logfire.msg",
    "value": {
      "stringValue": "chat gpt-5-2025-08-07",
    },
  },
  {
    "key": "logfire.json_schema",
    "value": {
      "stringValue": "{"type":"object","properties":{"gen_ai.operation.name":{"type":"string"},"gen_ai.request.model":{"type":"string"},"gen_ai.system":{"type":"string"},"gen_ai.request.max_tokens":{},"gen_ai.response.finish_reasons":{},"gen_ai.input.messages":{},"gen_ai.output.messages":{},"http.response.status_code":{"type":"number"},"http.request.body.text":{"type":"string"},"http.response.body.text":{"type":"string"},"gen_ai.response.model":{"type":"string"},"gen_ai.response.id":{"type":"string"},"gen_ai.usage.input_tokens":{"type":"number"},"gen_ai.usage.cache_read_tokens":{"type":"number"},"gen_ai.usage.cache_write_tokens":{},"gen_ai.usage.output_tokens":{"type":"number"},"gen_ai.usage.input_audio_tokens":{},"gen_ai.usage.cache_audio_read_tokens":{},"gen_ai.usage.output_audio_tokens":{"type":"number"},"events":{}}}",
    },
  },
  {
    "key": "logfire.level_num",
    "value": {
      "intValue": 9,
    },
  },
  {
    "key": "gen_ai.operation.name",
    "value": {
      "stringValue": "chat",
    },
  },
  {
    "key": "gen_ai.request.model",
    "value": {
      "stringValue": "gpt-5",
    },
  },
  {
    "key": "gen_ai.system",
    "value": {
      "stringValue": "openai",
    },
  },
  {
    "key": "gen_ai.request.max_tokens",
    "value": {},
  },
  {
    "key": "gen_ai.response.finish_reasons",
    "value": {},
  },
  {
    "key": "gen_ai.input.messages",
    "value": {},
  },
  {
    "key": "gen_ai.output.messages",
    "value": {},
  },
  {
    "key": "http.response.status_code",
    "value": {
      "intValue": 200,
    },
  },
  {
    "key": "http.request.body.text",
    "value": {
      "stringValue": "{
  "model": "gpt-5",
  "messages": [
    {
      "role": "developer",
      "content": "You are a helpful assistant."
    },
    {
      "role": "user",
      "content": "What is the capital of France?"
    }
  ]
}",
    },
  },
  {
    "key": "http.response.body.text",
    "value": {
      "stringValue": "{"id":"chatcmpl-CEY7M6BwbBtavblK5zBwrXgbDdIaF","object":"chat.completion","created":1757583776,"model":"gpt-5-2025-08-07","choices":[{"index":0,"message":{"role":"assistant","content":"Paris.","refusal":null,"annotations":[]},"finish_reason":"stop"}],"usage":{"prompt_tokens":23,"completion_tokens":11,"total_tokens":34,"prompt_tokens_details":{"cached_tokens":0,"audio_tokens":0},"completion_tokens_details":{"reasoning_tokens":0,"audio_tokens":0,"accepted_prediction_tokens":0,"rejected_prediction_tokens":0},"pydantic_ai_gateway":{"cost_estimate":0.00013875}},"service_tier":"default","system_fingerprint":null}",
    },
  },
  {
    "key": "gen_ai.response.model",
    "value": {
      "stringValue": "gpt-5-2025-08-07",
    },
  },
  {
    "key": "gen_ai.response.id",
    "value": {
      "stringValue": "chatcmpl-CEY7M6BwbBtavblK5zBwrXgbDdIaF",
    },
  },
  {
    "key": "gen_ai.usage.input_tokens",
    "value": {
      "intValue": 23,
    },
  },
  {
    "key": "gen_ai.usage.cache_read_tokens",
    "value": {
      "intValue": 0,
    },
  },
  {
    "key": "gen_ai.usage.cache_write_tokens",
    "value": {},
  },
  {
    "key": "gen_ai.usage.output_tokens",
    "value": {
      "intValue": 11,
    },
  },
  {
    "key": "gen_ai.usage.input_audio_tokens",
    "value": {},
  },
  {
    "key": "gen_ai.usage.cache_audio_read_tokens",
    "value": {},
  },
  {
    "key": "gen_ai.usage.output_audio_tokens",
    "value": {
      "intValue": 0,
    },
  },
  {
    "key": "events",
    "value": {
      "arrayValue": {
        "values": [],
      },
    },
  },
]
`;
